[Feature] params.require accepts array of parameters that should be present or raise error

gaurish opened the pull request:
This PR adds a method `require_all` which allows you to require multiple values in one method.

for example, in our app. I have a person object which MUST have first_name & last_name. so to do this in current code, I have to write this:

``` ruby
params.require(:person).require(:first_name)
params.require(:person).require(:last_name)
```

Here it will be one line for each params, so say if I require 10params, it will be 10lines of repeated code which is not DRY. So I propose a method which does this in one line:

``` ruby
params.require(:person).require_all(:first_name, :last_name)
```

Comments welcome

gaurish committed:
[Feature] params.require requires array of params

This PR adds ability to accept arrays which allows you to require multiple values in one method. so instead of this:

```ruby
params.require(:person).require(:first_name)
params.require(:person).require(:last_name)
```

Here it will be one line for each params, so say if I require 10params, it will be 10lines of repeated code which is not dry. So I have added new method which does this in one line:

```ruby
params.require(:person).require([:first_name, :last_name])
```

Comments welcome

egilburg commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -252,6 +252,10 @@ def require(key)
     # Alias of #require.
     alias :required :require
 
+    def require_all(keys)
```
Why not just make `def require(*keys)` and loop through them, instead of adding another mehod.

gaurish replied:
If that's more preferable I can do that too. As its not important to me if its one method or two methods. I just want to have an ability to require multiple params in single call.

so let me know what is more preferable. incase case its `def require(*keys)`, how do you handle the return value with multiple fields?

egilburg replied:
Not sure actually, but the same problem would then apply to `require_all` as well.

Typically the pattern is to use `require` for parent hashes, and `permit` for individual attributes. An individual attribute requirement is usually checked by model validation.

gaurish replied:
Validations work well if you want to store a value. However, if you want to do some pre-processing on param before storing it, you have to add `before_validation` action which works smaller apps. But incase of larger apps where same model is being used in multiple controllers(Web vs API controllers for example), it gets you into conditional callbacks which are not exactly clean. 

This is where ability to guard against `Nil` at controller level  is simple, clean way where what params are needed are documented in controller itself. 

Now, I am curious to know what are the cons of this [approach](http://stackoverflow.com/questions/22487878/strong-parameters-require-multiple) according to you.


if you think this idea is good, then I will improve the code & try to address issues you have mentioned.

egilburg replied:
The issue is that in Strong Parameters, `permit` is the heart of the logic that marks the attribute as whitelisted for future use in models. `require` simply will raise if not provided value, and AFAIK is mostly a glorified `presence` check before the `permit`.

If using `require` on individual attributes, you'd *still* need to `permit` them as well, or they will not be accepted by model (the most common use case), and `require` doesn't automatically `permit` the attribute.

arthurnn commented:
We have a similar use cases at Shopify.

thoughts on the API @dhh @rafaelfranca

dhh commented:
I like #require(*keys). If you ask for one key, you just get that value directly. If you ask for multiple keys, you get an array of values. This of course means that params.require(:two, :things).permit won't work, since permit only works on a hash. But imo that's fine.

egilburg commented:
What about require(one: { :thing1, :thing2 }) syntax? Then we can combine it with permit (by returning a hash-like object that responds to permit):

require(one: { :thing1, :thing2 }).permit(:thing3)

It's up to debate whether using this syntax should implicitly permit thing1 and thing2 to avoid having to repeat them in the permit part, but technical implementation could go either way.

Eugene

> On Mar 30, 2015, at 8:39 AM, David Heinemeier Hansson <notifications@github.com> wrote:
> 
> I like #require(*keys). If you ask for one key, you just get that value directly. If you ask for multiple keys, you get an array of values. This of course means that params.require(:two, :things).permit won't work, since permit only works on a hash. But imo that's fine.
> 
> —
> Reply to this email directly or view it on GitHub.
>

egilburg commented:
Typo, meant require(one: [ :thing1, :thing2 ]).permit(:thing3)

For single value could also accept shortcut syntax of require(one: :thing1).permit(:thing2)

Eugene

> On Mar 30, 2015, at 9:11 AM, Eugene Gilburg <eugene.gilburg@gmail.com> wrote:
> 
> What about require(one: { :thing1, :thing2 }) syntax? Then we can combine it with permit (by returning a hash-like object that responds to permit):
> 
> require(one: { :thing1, :thing2 }).permit(:thing3)
> 
> It's up to debate whether using this syntax should implicitly permit thing1 and thing2 to avoid having to repeat them in the permit part, but technical implementation could go either way.
> 
> Eugene
> 
>> On Mar 30, 2015, at 8:39 AM, David Heinemeier Hansson <notifications@github.com> wrote:
>> 
>> I like #require(*keys). If you ask for one key, you just get that value directly. If you ask for multiple keys, you get an array of values. This of course means that params.require(:two, :things).permit won't work, since permit only works on a hash. But imo that's fine.
>> 
>> —
>> Reply to this email directly or view it on GitHub.
>>

dhh commented:
I don't want to mix require and permit semantics. Require means return the
value, raise if its not there. Permit means, return key in a hash, no
raising.

On Mon, Mar 30, 2015 at 6:16 PM, Eugene Gilburg <notifications@github.com>
wrote:

> Typo, meant require(one: [ :thing1, :thing2 ]).permit(:thing3)
>
> For single value could also accept shortcut syntax of require(one:
> :thing1).permit(:thing2)
>
> Eugene
>
>
> > On Mar 30, 2015, at 9:11 AM, Eugene Gilburg <eugene.gilburg@gmail.com>
> wrote:
> >
> > What about require(one: { :thing1, :thing2 }) syntax? Then we can
> combine it with permit (by returning a hash-like object that responds to
> permit):
> >
> > require(one: { :thing1, :thing2 }).permit(:thing3)
> >
> > It's up to debate whether using this syntax should implicitly permit
> thing1 and thing2 to avoid having to repeat them in the permit part, but
> technical implementation could go either way.
> >
> > Eugene
> >
> >> On Mar 30, 2015, at 8:39 AM, David Heinemeier Hansson <
> notifications@github.com> wrote:
> >>
> >> I like #require(*keys). If you ask for one key, you just get that value
> directly. If you ask for multiple keys, you get an array of values. This of
> course means that params.require(:two, :things).permit won't work, since
> permit only works on a hash. But imo that's fine.
> >>
> >> —
> >> Reply to this email directly or view it on GitHub.
> >>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/19565#issuecomment-87739066>.
>

chancancode commented:
> I like #require(*keys). If you ask for one key, you just get that value directly. If you ask for multiple keys, you get an array of values.

I don't know if I am reading it right, is it something along the lines of...

```ruby
def require(*keys)
  if keys.length == 1
    # return value directly
  else
    # return array of values
  end
end
```

If that's the case, then I have a concern. The problem is you can't distinguish between "I'm asking for a single value" or "I'm passing an array that happens to contain only a single value". This makes it very annoying to use if you need to pass arrays with unknown number of elements:

```ruby
params.require( *REQUIRED_KEYS ) # might return a single value, or perhaps an array
```

This is a problem when you extract these things into constants, or when you moving these code into marcos/helpers that you reuse across controllers, or when writing library code. I'd much prefer:

```ruby
def require(keys)
  if Array === keys
    keys.map &method(:require)
  else
    # return single value
  end
end
```

So you'd have to call it by either `params.require( :thing )` or `params.require( [ :multiple, :things ] )` (or make those two separate methods as in the original proposal).

dhh commented:
Happy with the latter options too. IMO, require(array) is the minority
concern.

On Mon, Mar 30, 2015 at 7:18 PM, Godfrey Chan <notifications@github.com>
wrote:

> I like #require(*keys). If you ask for one key, you just get that value
> directly. If you ask for multiple keys, you get an array of values.
>
> I don't know if I am reading it right, is it something along the lines
> of...
>
> def require(*keys)
>   if keys.length == 1
>     # return value directly
>   else
>     # return array of values
>   endend
>
> If that's the case, then I have a concern. The problem is you can't
> distinguish between "I'm asking for a single value" or "I'm passing an
> array that happens to contain only a single value". This makes it very
> annoying to use if you need to pass arrays with unknown number of elements:
>
> params.require( *REQUIRED_KEYS ) # might return a single value, or perhaps an array
>
> This is a problem when you extract these things into constants, or when
> you moving these code into marcos/helpers that you reuse across
> controllers, or when writing library code. I'd much prefer:
>
> def require(keys)
>   if Array === keys
>     keys.map &method(:require)
>   else
>     # return single value
>   endend
>
> So you'd have to call it by either params.require( :thing ) or params.require(
> [ :multiple, :things ] ) (or make those two separate methods as in the
> original proposal).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/19565#issuecomment-87756667>.
>

chancancode commented:
Agreed. `params.require( :thing )` is probably still the most popular thing you need to do, so that should get the nice API regardless of what we pick for the array case.

In theory, the array thing allows you to write smart things like this:

```ruby
first_name, last_name = params.require([ :first_name, :last_name ])
```

But I don't think you can find too many excuses to use that in practice :)

litch commented:
I actually just implemented something similar in a project of mine and have been debating where to generalize it to.  The issue that I see is that this is really a mixing of concerns.  If you're doing parameter validation (which is what require_all, and REALLY what require is as well, then you should be able to do type checks).  The calling semantics I would reach for next after being able to do `require_all` would be:

```ruby
first_name, last_name, addresses = params.require( first_name: String, last_name: String, addresses: Array)
```

But that obviously turns into a mess in a hurry, and doesn't seem to me like it should be in strong params.

`permit` and `require` just seem like such different concerns to me, and start down such different paths, that it seems like it would be a bad idea to add this.

dhh commented:
Yup, fine with that extension. Agree it's the minority case.

On Mon, Mar 30, 2015 at 7:26 PM, Godfrey Chan <notifications@github.com>
wrote:

> Agreed. params.require( :thing ) is probably still the most popular thing
> you need to do, so that should get the nice API regardless of what we pick
> for the array case.
>
> In theory, the array thing allows you to write smart things like this:
>
> first_name, last_name = params.require([ :first_name, :last_name ])
>
> But I don't think you can find too many excuses to use that in practice :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/19565#issuecomment-87760600>.
>

egilburg commented:
```ruby
def require(thing, *things)
  things += thing if things.any?

  ...
end
```

A bit clunkier implementation but allows mainitaing the syntax.

dhh commented:
I'd rather be explicit. If you want an array of values, supply an array of keys. 


> On Mar 30, 2015, at 19:36, Eugene Gilburg <notifications@github.com> wrote:
> 
> def require(thing, *things)
>   things += thing if things.any?
> 
>   ...
> end
> A bit clunkier implementation but allows mainitaing the syntax.
> 
> —
> Reply to this email directly or view it on GitHub.
>

senny commented:
/cc @fxn

fxn commented:
I think there are two design drivers in conflict here.

In my mind, strong parameters is *params whitelisting for security purposes*. Applications declare what they accept, and strong parameters discards what doesn't match. Support for permitted scalars was motivated by past issues in which a hash could be injected where a string was assumed.

Then we have a second pulling force: request validation. This overlaps with model validation in most use cases of idiomatic Rails applications. For me, strong parameters doesn't address this problem, and in that sense `require` has always felt a little off to me.

Why add support to declare multiple required keys? Then, what about arbitrarily nested required keys? Presence is just one of many possible validations, what about limits on array size? Regexps for strings? Ranges for integers? Size for file uploads?

Personally, I wouldn't go in the direction of this PR because strong parameters is not about request validation.

dhh commented:
They are indeed separate concerns, but they're composing a whole that means
"get these params ready for model consumption". Whitelisting is the primary
element of that, but requiring feels like a reasonable wheel to go along
with particularly because it deals not with a business rule about what the
param contains, but merely "is it here or not".

The alternative to having require is to do params[:person].permit(:name,
:age) which will blow up when params[:person] isn't present. Nobody wants
to see their exception log filled with stuff like that.

So I like requiring is a reasonable 3rd wheel here as it pertains to
getting the params ready for #permit.

But I do think it's entirely fair to question whether #require makes sense
when it's NOT a pre-filter to a #permit call.

On Mon, Apr 6, 2015 at 3:15 PM, Xavier Noria <notifications@github.com>
wrote:

> I think there are two design drivers in conflict here.
>
> In my mind, strong parameters is *params whitelisting for security
> purposes*. Applications declare what they accept, and strong parameters
> discards what doesn't match. Support for permitted scalars was motivated by
> past issues in which a hash could be injected where a string was assumed.
>
> Then we have a second pulling force: request validation. This overlaps
> with model validation in most use cases of idiomatic Rails applications.
> For me, strong parameters doesn't address this problem, and in that sense
> require has always felt a little off to me.
>
> Why add support to declare multiple required keys? Then, what about
> arbitrarily nested required keys? Presence is just one of many possible
> validations, what about limits on array size? Regexps for strings? Ranges
> for integers? Size for file uploads?
>
> Personally, I wouldn't go in the direction of this PR because strong
> parameters is not about request validation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/19565#issuecomment-90059652>.
>

fxn commented:
For me, an API that would keep things at the same level and protect `permit` calls is something along this line:

```ruby
def action
  @person = Person.create(safe_params[:person])
end

def safe_params
  params.permit(person: [:name, :age])
end
```

Addressing buggy client code or hacked requests would need a new API I believe, one that validates that a params structure conforms to some specification.

dhh commented:
permit doesn't mutate the hash, which is what we want. Things enter as `{
person: { name: 'david', age: '35' }` but Person.create wants `{ name:
'david', age: '35' }`. I don't think we should change #permit to change the
structure of the hash.

On Mon, Apr 6, 2015 at 3:53 PM, Xavier Noria <notifications@github.com>
wrote:

> For me, an API that would keep things at the same level and protect permit
> calls is something along this line:
>
> def action
>   @person = Person.create(safe_params[:person])end
> def safe_params
>   params.permit(person: [:name, :age])end
>
> Addressing buggy client code or hacked requests would need a new API I
> believe, one that validates that a params structure conforms to some
> specification.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/19565#issuecomment-90071216>.
>

dhh commented:
In any case, I do think it fits under the charter of Strong Params to deal
with parameter validation to some extent. We don't want it to deal with
business rules, but type and presence constraints seem legit. Kinda like
how we let the database enforce unique constraints in some cases, but any
advanced logic lives in the model.

On Mon, Apr 6, 2015 at 3:56 PM, David Heinemeier Hannson <
david@loudthinking.com> wrote:

> permit doesn't mutate the hash, which is what we want. Things enter as `{
> person: { name: 'david', age: '35' }` but Person.create wants `{ name:
> 'david', age: '35' }`. I don't think we should change #permit to change the
> structure of the hash.
>
> On Mon, Apr 6, 2015 at 3:53 PM, Xavier Noria <notifications@github.com>
> wrote:
>
>> For me, an API that would keep things at the same level and protect
>> permit calls is something along this line:
>>
>> def action
>>   @person = Person.create(safe_params[:person])end
>> def safe_params
>>   params.permit(person: [:name, :age])end
>>
>> Addressing buggy client code or hacked requests would need a new API I
>> believe, one that validates that a params structure conforms to some
>> specification.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/rails/rails/pull/19565#issuecomment-90071216>.
>>
>
>

fxn commented:
Yeah, take into account in the previous example the action uses `safe_params[:person]`.

Also, conceptually, for me request validation is not really about business rules (although it is of course related to them). In my mind the server publishes a spec, client code must conform to that spec, and the server validates that an API request conforms to said spec, otherwise it is not passed down. That's for me request validation, and it is something strong parameters does not do as of today.

Anyway, that's how I'd evolve (or not evolve) strong parameters and it's fine of course that we have different points of view. I understand the approach of doing just that little thing, and I being able to declare nested required keys (for only one root key) it's a normal extension to the current API `require`.

fxn commented:
One more thing: being able to say that the first name of a person is required is unrelated to `permit`. If you do that in the controller it leans on the side of plain request validation, it is what typically has been validated at the model level, among the other many rules you can't validate in the controller anyway.

That's what makes me skeptical about this PR, we don't do request validation, but I have the intuition that a bit of that is being slipped here.

fxn commented:
Hey, since there has been a few messages today people in the thread may be wondering what's next :).

David's POV is clear, and strong parameters is his baby. I gave my feedback since I was cc'ed, but probably we won't agree, and that's fine. Let's move on with the PR!

dhh commented:
Very much appreciate the input, Xavier. And I think we should consider how
we can also deal with the parameter validation aspect through a suitable
API. We used to have that on the controller. Forgot what it was called.

On Mon, Apr 6, 2015 at 6:08 PM, Xavier Noria <notifications@github.com>
wrote:

> Hey, since there has been a few messages today people in the thread may be
> wondering what's next :).
>
> David's POV is clear, and strong parameters is his baby. I gave my
> feedback since I was cc'ed, but probably we won't agree, and that's fine.
> Let's move on with the PR!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/19565#issuecomment-90126369>.
>

gaurish commented:
So what needs to be done here, any pointers?

gaurish commented:
Updated PR based on the suggestions. To summarise the discussion.

1. We need request validation(just type & presence) on controller level.
2. I have updated the PR, removing `require_all` & added support for `require` to accept arrays. 

so you can do 

```ruby
first_name, title = params.require([ :first_name, :title ])
```

if `first_name` or `title` params are empty, it will generate an error back to client with 400 bad request -- client, allowing client errors to be really be client errors & not server errors(5xx)


@dhh 
Any thoughts, Feedback?

dhh commented:
API looks good to me. Someone else might review the implementation but also looks pretty straight forward to me.

kaspth commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -253,11 +253,11 @@ def permit!
-      if value.present? || value == false
-        value
+      if key.is_a?(Array)
+        keys.map &method(:require)
```
It would be an easier change if we just do an early return:

```ruby
def require(key)
  return key.map(&method(:require)) if key.is_a?(Array)

  # Rest of the method doesn't need to change at all.
end
```

kaspth replied:
Also lets create a method object, and just use the standard map invocation. That's 80% faster:

```ruby
require 'benchmark/ips'

def hello(key)
  key.upcase
end

keys = %i(a b c d e f)

Benchmark.ips do |x|
  x.report('method')   { keys.map(&method(:hello)) }
  x.report('standard') { keys.map { |k| hello(k) } }

  x.compare!
end
```

Results:

```
Calculating -------------------------------------
              method    22.731k i/100ms
            standard    37.384k i/100ms
-------------------------------------------------
              method    267.768k (± 2.4%) i/s -      1.341M
            standard    483.950k (± 2.9%) i/s -      2.430M

Comparison:
            standard:   483950.4 i/s
              method:   267768.2 i/s - 1.81x slower
```

kaspth commented on actionpack/test/controller/required_params_test.rb:
```diff
@@ -48,4 +48,20 @@ class ParametersRequireTest < ActiveSupport::TestCase
+      .require(:person)
+      .require([:first_name, :last_name])
+      assert_kind_of Array, safe_params
+      assert_equal ['Gaurish', 'Mjallo'], safe_params
```
The asserts are indented by two spaces more than they should. Let's also add a new line after the last require.

gaurish commented:
Updated as per Feedback.

kaspth commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -252,13 +252,17 @@ def permit!
-      else
-        raise ParameterMissing.new(key)
-      end
+      (value.present? || value == false) ? value : raise(ParameterMissing.new(key))
```
My point was that these lines shouldn't be changed at all. Can you roll back to how it was before your change?

gaurish replied:
Done.

kaspth commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -252,13 +252,17 @@ def permit!
+    #   first_name, title = params.require([:first_name, :title])
+    #
     def require(key)
+      return keys.map{|k| require(k)} if key.is_a?(Array)
```
Add spaces around the map like this: `map { |k| require(k) }`

kaspth commented:
Thanks

gaurish commented:
Thanks for reviewing 😉

sikachu commented:
I know I'm late to the party, but I have a feeling that this promotes a feature to be used in the way that it's not intended to be.

What about actually adding [`assert_valid_keys`](http://api.rubyonrails.org/classes/Hash.html#method-i-assert_valid_keys) to ActionController::Parameters? We already have that on Hash, and it's currently the way to make sure a Hash has all the given keys. Adding this recursive functionality to `require` makes this method doing too much IMHO.

fxn commented:
The apparently innocent patch had an important typo that wasn't catched by the tests, when through it in cbe7899f9d5fece4749f75828fd120d67056f356.

If correcting a patch about an API you don't agree with isn't team playing I don't know what it is! :smile:

fxn commented:
The examples are not good for certain, it is not obvious to the reader that `first_name` could be an injected hash because it has not been permitted.

Maybe if we had a non-artificial use case for an action that receives several root objects... anyone?

fxn commented:
I can't think of a use case. Say

```ruby
def user_params
  params.require(:user).permit(:name, :email, :bio).tap do |up|
    up.require([:name, :email])
  end
end
```

Besides being really ugly for my taste, are we implementing a presence validator in the controller? Why? That's tipically a concern of the model (together with several other validators not available anyway). We have never promoted using `require` that way AFAIK.

On the other hand `params.require(:user)` is not a data validator, it is asserting the action expects a form for users, which is different (and as I argued above, still feels a bit off to me, but that has become idiomatic).

Regarding several root users, I cannot find a nice example:

```ruby
def safe_params
    user, product = params.require([:user, :product])
    return user.permit(...), product.permit(...) 
end
```

versus

```ruby
def user_params
  # ...
end

def product_params
  # ...
end
```

Which usage patterns should Rails promote with this API?

fxn commented:
@sikachu @dhh let me make something explicit to make sure we are all talking about the same.

`require` is a *value* presence validator, not a *key* presence validator. If a form sends the user email as a blank field, and the controller checks `require(:email)`, then we get a parameter missing exception albeit the parameter was sent.

I am pretty sure this was understood in the discussion above, I don't think going in that direction is good for strong params, but it was fine for David (and I agreed to disagree, so not discussing the acceptance of the PR, only following up because we need non-artificial examples for the docs, because the one in the patch can't end up there).

And if you think about it, this discussion is really unrelated to this PR, it is about promoting the use of `require` for scalars as a presence validator vs promoting only its usage for root objects. This same discussion could have happened without support for arrays.

dhh commented:
Right. I had a couple of cases in my own code base with forms that didn't
have a shared root. Like user_name and password instead of
login[:user_name] and login[:password]. Using value presence checking on
that case made perfect sense to me.

On Sat, Aug 29, 2015 at 3:04 AM, Xavier Noria <notifications@github.com>
wrote:

> @sikachu <https://github.com/sikachu> @dhh <https://github.com/dhh> let
> me make something explicit to make sure we are all talking about the same.
>
> require is a *value* presence validator, not a *key* presence validator.
> If a form sends the user email as a blank field, and the controller checks
> require(:email), then we get a parameter missing exception albeit the
> parameter was sent.
>
> I am pretty sure this was understood in the discussion above, I don't
> think going in that direction is good for strong params, but it was fine
> for David (and I agreed to disagree, so not discussing the acceptance of
> the PR, only following up because we need non-artificial examples for the
> docs, because the one in the patch can't end up there).
>
> And if you think about it, this discussion is really unrelated to this PR,
> it is about promoting the use of require for scalars as a presence
> validator vs promoting only its usage for root objects. This same
> discussion could have happened without support for arrays.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/19565#issuecomment-135959727>.
>
